src/main/java/org/traccar/helper/model/PositionUtil.java

```diff
@@ -64,14 +64,17 @@ public static List<Position> getPositions(
                 new Order("fixTime")));
     }
 
-    public static List<Position> getLatestPositions(Storage storage, long userId) throws StorageException {
-        var devices = storage.getObjects(Device.class, new Request(
-                new Columns.Include("id"),
-                new Condition.Permission(User.class, userId, Device.class)));
+    public static List<Position> getLatestPositions(Storage storage, long userId, List<Device> devices)
+            throws StorageException {
+        if (devices == null) {
```

Comment:
I think the right solution is to do a single query that would combine devices query with positions.

Reply:
I agree. I’ll refactor.

Reply:
While working on this suggestion, I found that `org.traccar.storage.DatabaseStorage#getObjects` only instantiates objects of a given type, and the members should match the SQL ResultSet. In the case of joining Positions to Devices in the Devices query, I would only be able to instantiate one of the other, unless I created another type of object (something akin to DevicePosition).

Would this be your recommendation?

Reply:
I'm not proposing to join tables, but filter positions by devices.

Reply:
From your first message in this thread proposing a _single query that would combine devices query with positions_ I see no other way to do this than performing a join...

Here would be the proposed query which would seem to work:

```sql
SELECT * FROM tc_devices td
INNER JOIN tc_positions tp on tp.id = td.positionid 
WHERE td.id IN (SELECT tc_user_device.deviceId FROM tc_user_device WHERE userId = ? UNION SELECT DISTINCT deviceId FROM tc_user_group INNER JOIN (SELECT id as parentId, id as groupId FROM tc_groups UNION SELECT groupId as parentId, id as groupId FROM tc_groups WHERE groupId IS NOT NULL UNION SELECT g2.groupId as parentId, g1.id as groupId FROM tc_groups AS g2 INNER JOIN tc_groups AS g1 ON g2.id = g1.groupId WHERE g2.groupId IS NOT NULL) AS all_groups ON tc_user_group.groupId = all_groups.parentId INNER JOIN (SELECT groupId as parentId, id as deviceId FROM tc_devices WHERE groupId IS NOT NULL) AS devices ON all_groups.groupId = devices.parentId WHERE userId = ?);
```

---

But as mentioned the Traccar DB interface in its current form only allows returning a single type of object.

When you suggest, however, filtering positions by devices, that seems to be exactly what I'm doing with the IN operator in this PR. Your current implementation also filters positions by devices, but it does this only after creating Java objects from all of the latest positions, flooding the heap when the amount of devices in the DB is large (>40k).

Would love to implement this in a way the project would appreciate. 😃

Reply:
I don't understand why you need joins. What I'm proposing is something like this:
```sql
SELECT * FROM tc_positions WHERE id IN (
  SELECT positionId FROM tc_devices WHERE id IN (
    ... something similar to the current permission query
  )
)

Reply:
Thanks for the clarification Anton. I understand now.

However, this again adds the issue that we can't reuse the list of devices accessible to the user and this is the most expensive part of the servlet request right now. In all of my measurements the permission query takes the majority (>80%) of the wall time. I already refactored so it would only happen once. If we cannot save the deviceIds from the permission query then we repeat it again which doubles the wall time requirement of the servlet request again.

See attached what the servlet request looks like before removing the second call. Your proposal would require the second call again.

![image](https://github.com/traccar/traccar/assets/5489134/80309d8c-cf2e-4d62-89e8-80287ae611af)

Reply:
That's a different issue and I think it's a reasonable trade-off.